### ISSUE TITLE
Fine-tune ViT with LoRA

### ISSUE THREAD

**Issue by jedrzejwalega** [Appsilon/image_flow_cytometry_fine_tune#8]
Now that we've [recreated](https://github.com/orgs/Appsilon/projects/154/views/1?pane=issue&itemId=87598804&issue=Appsilon%7Cimage_flow_cytometry_fine_tune%7C4) the authors' ResNet18 results (mostly to confirm the lightning and neptune framework works same way as their skorch) we want to try more recent models than ResNet.

The goal is to fine-tune ConvNext (pretrained can be taken from timm) on our [dataset](https://console.cloud.google.com/storage/browser/image-flow-cytometry-finetune/metadata?inv=1&invt=AbiAZQ&project=marekresearch&pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))).

We are benchmarking against the authors' val F1 = 0.91.

We would like to use LoRA for this. The choice of ViT is not the most important part in this experiment (we could try it with some other nn), but we want to check whether LoRA would be suitable for our modest (2.9k train) set of images available.

I'm going to learn `perf` library to apply LoRA onto ViT.

**Comment by jedrzejwalega** [comment]
- Fine-tuning [code](https://github.com/Appsilon/image_flow_cytometry_fine_tune/tree/finetune/6%20-%20Appsilon/1%20-%20Machine%20learning/Fine-tuning%20ViT%20LoRA)
- Model [implementation wit LoRA](https://github.com/Appsilon/image_flow_cytometry_fine_tune/blob/finetune/6%20-%20Appsilon/utils/vit_lora.py)

**Comment by jedrzejwalega** [comment]
First, short [run](https://app.neptune.ai/o/appsilon/org/image-flow-cytometry-finetune/runs/details?viewId=standard-view&detailsTab=charts&shortId=IM-25&type=run) with `learning_rate = 0.0004`, same as the one we use for ConvNext.

Doesn't work so well. Discussed with @Peterdes. I'm going to apply a few [enhancements](https://github.com/Appsilon/image_flow_cytometry_fine_tune/commit/d3c0d23944a4ba2337cebc11683a6f67c2a96803) to my code:
- Reducing LoRA rank to 8 as the original [paper](https://arxiv.org/pdf/2106.09685) recommends for models of this size
- [Turning on](https://github.com/Appsilon/image_flow_cytometry_fine_tune/commit/d1892461bda02932709370f4b9ea0d61ff781644) rslora ([explained](https://huggingface.co/blog/damjan-k/rslora))
- In my LoRA implementation I've applied it to not just all attention weights, but also intermediate layers and output one. That might be too much, so I'll go with just [query and value](https://github.com/Appsilon/image_flow_cytometry_fine_tune/commit/3a9d6ae0fc9b4a8ee8a14485025bc2f04f29f10e). I don't want to freeze especially the output layer, as it's getting replaced by us providing a custom number of classification classes, so it needs to train from scratch.

**Comment by jedrzejwalega** [comment]
[10 epoch run](https://app.neptune.ai/o/appsilon/org/image-flow-cytometry-finetune/runs/details?viewId=standard-view&detailsTab=charts&shortId=IM-35&type=run) seemed too short to conclude, I'm letting it run for [50](https://app.neptune.ai/o/appsilon/org/image-flow-cytometry-finetune/runs/details?viewId=standard-view&detailsTab=charts&shortId=IM-37&type=run).

**Comment by jedrzejwalega** [comment]
[Longer run](https://app.neptune.ai/o/appsilon/org/image-flow-cytometry-finetune/runs/details?viewId=standard-view&detailsTab=dashboard&dashboardId=IFC-9d8b15a3-7a6b-4d6b-98e4-606ee4c3ad81&shortId=IM-47&type=run) does not exceed validation F1 = 0.7.

It's a far shot from we're aiming for (val F1 > 0.91). Thinking of why, I'm suspecting that regular fine tuning of classification and patch embedding layers should be separated from the LoRA fine-tuning of the query-value attention blocks. At the moment it is performed simultaneously.
![Image](https://github.com/user-attachments/assets/229d250a-46c3-4570-aa94-25fdf00d2ae7)

Around epoch 9 (step ~1000) we can see a big drop in accuracy, reflecting, most likely, the training of the patch embedding layer. 

While the LoRA separation seems plausible for me, there are multiple other factors to consider as to why the approach doesn't work as expected. First off - the augmentations, defined by the authors, seem weird and non-exhaustive. I haven't touched them so far to minimize interference in the already established benchmark, but perhaps this isn't the way.

After internal discussion we have decided to stop the current strategy - simply modifying the authors' approach and start off a brand new route with a from-scratch solution.
